### PR TITLE
fix(hax-lib/int): make `Int` division, `rem_euclid` and `pow2` sound

### PR DESCRIPTION
**Pull request by abentkamp**
This PR fixes the F* and Lean models of hax_lib's `Int`.

Unfortunately, the models are still not perfect because hax_lib's `Int` is limited to `1024` bytes. When it overflows, it panics. I created this issue for it: https://github.com/cryspen/hax/issues/2260

I started to use the new Changelog format we discussed. Are soundness fixes supposed to go into "Security"?

Assisted-by: Claude Opus 5

### CODE REVIEW

**Comment by abentkamp** [comment]
Can we include the HaxLib models into our core model testing strategy somehow to prevent these kinds of soundness issues?

File: engine/backends/fstar/fstar_backend.ml [hunk 1]
```diff
@@ -329,7 +329,9 @@ struct
       (c Rust_primitives__hax__int__add, (2, "+"));
       (c Rust_primitives__hax__int__sub, (2, "-"));
       (c Rust_primitives__hax__int__mul, (2, "*"));
-      (c Rust_primitives__hax__int__div, (2, "/"));
+      (* `Int` division has no infix form here: it truncates towards zero,
+         while F*'s `/` on `int` is Euclidean. It prints as a call to
+         `Rust_primitives.Hax.Int.div`, which spells out the difference. *)
       (c Rust_primitives__hax__int__rem, (2, "%"));
```
Comment (maximebuyse) [inline engine/backends/fstar/fstar_backend.ml]: It seems like there is no implementation of `Rem` for hax_lib integers so this line is dead. But if we add an impl, the same mismatch as for division would happen. It would make sense to delete it.

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,21 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+
+### Changed
+
+hax-lib crate:
+ - Allow any positive exponent for hax_lib::Int::pow2
```
Comment (maximebuyse) [inline CHANGELOG.md]: Not really any exponent because of the 1024 bytes limit

Comment (maximebuyse) [reply CHANGELOG.md]: Actually since 2^8191 is the biggest power of 2 fitting in 1024 bytes, and 8191 fits in a u32, I think your change doesn't allow more exponents than the previous version. The real improvement is that the error message is better on negative exponents

**Review by maximebuyse — COMMENTED** [review]
Not very critical but good to fix this. I think it should be possible to test this in the Rust/Lean equivalence tests, we'd need to add the hax_lib dependency there. And we'd be able to add tests for the division on negative numbers (that's already good to add), but testing that something panics requires some new infrastructure.

**Comment by maximebuyse** [comment]
Also, legacy-lean still has similar issues, we should at least document it.

File: engine/backends/fstar/fstar_backend.ml (hunk 1 above)
Comment (abentkamp) [reply engine/backends/fstar/fstar_backend.ml]: Removed.

File: CHANGELOG.md (hunk 2 above)
Comment (abentkamp) [reply CHANGELOG.md]: Right, this is barely a user-facing change. I removed that changelog entry entirely.

**Comment by abentkamp** [comment]
I added a fix for legacy lean as well. It doesn't model `rem_euclid` and `pow2` at all. I think that's okay; I'd suggest to fix soundness issues only for legacy-lean.

**Review by maximebuyse — APPROVED** [review]
